crates/network-scanner-net/src/async_io.rs

```diff
@@ -0,0 +1,152 @@
+use std::{
+    collections::HashMap,
+    num::NonZeroUsize,
+    sync::{
+        atomic::{AtomicBool, AtomicUsize},
+        Arc,
+    },
+};
+
+use anyhow::Context;
+use crossbeam::channel::Receiver;
+use parking_lot::Mutex;
+use polling::{Event, Events};
+use socket2::Socket;
+
+use crate::async_raw_socket::AsyncRawSocket;
+
+#[derive(Debug)]
+pub struct Socket2Runtime {
+    poller: polling::Poller,
+    next_socket_id: AtomicUsize,
+    is_terminated: AtomicBool,
+    map: Arc<Mutex<HashMap<usize, std::task::Waker>>>,
+    sender: crossbeam::channel::Sender<(Event, std::task::Waker, Arc<Socket>)>,
+}
+
+impl Drop for Socket2Runtime {
+    fn drop(&mut self) {
+        self.is_terminated.store(true, std::sync::atomic::Ordering::SeqCst);
+        self.poller
+            .notify()
+            .map_err(|e| tracing::error!("failed to notify poller: {:?}", e))
+            .ok();
+        // event loop will terminate after this
+        // register loop will terminate becase of sender is dropped after this.
+    }
+}
+
+impl Socket2Runtime {
+    pub fn new() -> anyhow::Result<Arc<Self>> {
+        let poller = polling::Poller::new()?;
+        //unbounded channel performance is worse, but prevents sender from blocking, in case of Tokio crashes completely
+        let (sender, receiver) = crossbeam::channel::unbounded();
```

Comment:
sorry, I did not explain clearly what was the concern. 
If the channel is full, then the sender would block and will crashes the thread that polls the socket's future.
So by setting it as unbounded, we made sure that the sending end of the channel will never block. 

Furthermore, since the the receiving end is running on a dedicated loop, it is not very likely to stack up the channel.
I agree that a sized channel might be better depending on the usage.

Reply:
updated